client/allocrunner/task_coordinator_test.go

```diff
@@ -0,0 +1,449 @@
+
+	// Init task is done, main is now allowed to run.
+	states = map[string]*structs.TaskState{
+		initTask.Name: {
+			State:  structs.TaskStateDead,
+			Failed: false,
+		},
+		sideTask.Name: {
+			State:  structs.TaskStateRunning,
+			Failed: false,
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskBlocked(t, coord, initTask)
+	requireTaskAllowed(t, coord, sideTask)
+	requireTaskAllowed(t, coord, mainTask)
+}
+
+func TestTaskCoordinator_MainRunsAfterManyInitTasks(t *testing.T) {
+	ci.Parallel(t)
+
+	logger := testlog.HCLogger(t)
+
+	alloc := mock.LifecycleAlloc()
+	alloc.Job = mock.VariableLifecycleJob(structs.Resources{CPU: 100, MemoryMB: 256}, 1, 2, 0)
+	tasks := alloc.Job.TaskGroups[0].Tasks
+
+	mainTask := tasks[0]
+	init1Task := tasks[1]
+	init2Task := tasks[2]
+
+	shutdownCh := make(chan struct{})
+	defer close(shutdownCh)
+	coord := newTaskCoordinator(logger, tasks, shutdownCh)
+
+	// All tasks start blocked.
+	requireTaskBlocked(t, coord, init1Task)
+	requireTaskBlocked(t, coord, init2Task)
+	requireTaskBlocked(t, coord, mainTask)
+
+	// Set initial state, prestart tasks are allowed to run, mais is blocked.
+	states := map[string]*structs.TaskState{
+		init1Task.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		init2Task.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, init1Task)
+	requireTaskAllowed(t, coord, init2Task)
+	requireTaskBlocked(t, coord, mainTask)
+
+	// Init tasks complete, main is allowed to run.
+	states = map[string]*structs.TaskState{
+		init1Task.Name: {
+			State:      structs.TaskStateDead,
+			Failed:     false,
+			StartedAt:  time.Now(),
+			FinishedAt: time.Now(),
+		},
+		init2Task.Name: {
+			State:     structs.TaskStateDead,
+			Failed:    false,
+			StartedAt: time.Now(),
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskBlocked(t, coord, init1Task)
+	requireTaskBlocked(t, coord, init2Task)
+	requireTaskAllowed(t, coord, mainTask)
+}
+
+func TestTaskCoordinator_FailedInitTask(t *testing.T) {
+	ci.Parallel(t)
+
+	logger := testlog.HCLogger(t)
+
+	alloc := mock.LifecycleAlloc()
+	alloc.Job = mock.VariableLifecycleJob(structs.Resources{CPU: 100, MemoryMB: 256}, 1, 2, 0)
+	tasks := alloc.Job.TaskGroups[0].Tasks
+
+	mainTask := tasks[0]
+	init1Task := tasks[1]
+	init2Task := tasks[2]
+
+	shutdownCh := make(chan struct{})
+	defer close(shutdownCh)
+	coord := newTaskCoordinator(logger, tasks, shutdownCh)
+
+	// All tasks start blocked.
+	requireTaskBlocked(t, coord, init1Task)
+	requireTaskBlocked(t, coord, init2Task)
+	requireTaskBlocked(t, coord, mainTask)
+
+	// Set initial state, prestart tasks are allowed to run, mais is blocked.
```

Comment:
(there's a few of these in this file copy-and-pasted)

```suggestion
	// Set initial state, prestart tasks are allowed to run, main is blocked.
```

Reply:
Fun fact: "mais" means "more" is Portuguese 😬